**Привет!**

Меня зовут Артемьева Саша, я буду проверять твой проект. Приятно познакомиться, хоть и заочно:)
    
Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Это как тестовое задание при приеме на работу: очень много мелких ошибок могут стать причиной отказа кандидату. 

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')

#запросим общую информацию о датасете и посмотрим как представлены данные
display(data.info()) 
display(data.head(10))

#посмотрим, какие уникальные значения встречаются в каждом столбце
display(data['children'].sort_values().unique())
display(data['days_employed'].sort_values().unique())
display(data['dob_years'].sort_values().unique())
display(data['education'].sort_values().unique())
display(data['education_id'].sort_values().unique())
display(data['family_status'].sort_values().unique())
display(data['family_status_id'].sort_values().unique())
display(data['gender'].sort_values().unique())
display(data['income_type'].sort_values().unique())
display(data['debt'].sort_values().unique())
display(data['total_income'].sort_values().unique())
display(data['purpose'].sort_values().unique())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

array([-1,  0,  1,  2,  3,  4,  5, 20])

array([-18388.94990057, -17615.56326563, -16593.47281726, ...,
       401715.81174889, 401755.40047533,             nan])

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

array([0, 1, 2, 3, 4])

array(['Не женат / не замужем', 'в разводе', 'вдовец / вдова',
       'гражданский брак', 'женат / замужем'], dtype=object)

array([0, 1, 2, 3, 4])

array(['F', 'M', 'XNA'], dtype=object)

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

array([0, 1])

array([  20667.26379327,   21205.28056622,   21367.64835649, ...,
       2200852.2102589 , 2265604.02872274,              nan])

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

**Вывод**

Данные в датасете содержат 12 столбцов и 21525 строк. 7 столбцов содержат количественные переменные, 5 столбцов - категориаальные переменные. В столбцах 'total_income', 'days_employed', 'gender' встречаются пропуски в данных.

В столбце 'gender' встречается значение 'XNA', которое, вероятно, связано с тем, что заемщик при заполнении анкеты не указал свой пол.

Также в столбце 'days_employed' встречаются отрицательные значения. И сами значения, если это действительно, дни - нереальны (стаж более чем в 100 лет и т.д.). Ошибка в форматировании данных? Стаж в днях указан в Unix time?

В столбце 'children' встречаются отрицательные значения (количество детей =-1).
В столбце 'dob_years' встречаются нули (возраст равен 0). Возможно, потому, что заемщик не указал возраст при заполнении анкеты.

Значения в столбцах 'purpose', 'education' дублируются (текстовые переменные).

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Увидели, с чем предстоит иметь дело

## Шаг 2. Предобработка данных

### Обработка пропусков

In [2]:
#проверим, как выглядят строки с пропусками в данных:
isna = pd.isna(data['total_income'])
display(data[isna])

#сравним среднее и медианное значения в столбце с уровнем дохода ('total_income')
display(data['total_income'].mean())
display(data['total_income'].median())

#среднее и медианное значение отличааются не столь существенно
#можно использовать среднее значение при заполнении пропусков в данных

#заполним пропуски в столбце 'total_income'
data['total_income'].fillna(value = data['total_income'].mean(), inplace = True)

#заменим пропуски в столбце о трудовом стаже на ноль (так как в столбце встречаются отрицатльные значения 
#и мы также не можем достоверно утверждать, по какому критерию нам надо изменить данные, чтобы получить верные значения)
data['days_employed'].fillna(value = 0, inplace = True)

#проверим, избавились ли мы от пропусков в данных
display(data.info())


children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

167422.30220817294

145017.93753253992

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

**Вывод**
Отмечено, что строки с пропусками в столбце 'total_income', как правило, также содержат пропуски в столбце 'days_employed'. Возможно, это связано с тем, что самим заемщиком при заполнении заявки не предоставлены сведения о трудовом стаже и доходе.






<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Можем сообщить о проблеме заказчику и по возможности запросить метаданные для дальнейшего анализа

### Замена типа данных

In [3]:
#заменим формат данных в столбцах 'days_employed' и 'total_income': 
#вместо вещественных переменных будем использовать целочисленные.
#Метод .astype используем, т.к. при использовании .to_numeric получаем как раз вещественную переменную
data['days_employed'] = data['days_employed'].astype('int')
#заменим отрицательные значения в столбце на их абсолютные значения методом .abs()
data['days_employed'] = data['days_employed'].abs() 
data['total_income'] = data['total_income'].astype('int')
data['children'] = data['children'].abs()#заменим отрицательные значения в столбце на их абсолютные значения




**Вывод**

Мы заменили формат значений на целочисленный в 2 столбцах: с трудовым стажем и с уровнем дохода, применив метод .astype
Чтобы обработать отрицательные значения из столбца, которые, вероятно, являются ошибочными, применили метод .abs(), заменяющий целочисленную переменную на ее модуль (абсолютное значение).

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Можно было бы также обработать странное значение пола и кол-во детей в 20 человек. Это явные аномалии. В целом, я бы советовала сразу высчитывать долю аномальных значений и если она не превышает 5%, не тратить время на их обработку, а смело от них избавляться

### Обработка дубликатов

In [4]:
#приведем значения в датафрейме к нижнему регистру  и проверим результат
data['education'] = data['education'].str.lower()
display(data['education'].sort_values().unique()) 

#подсчитаем число дублей в датафрейме
display(data.duplicated().sum())

#удаляем дубликаты
data = data.drop_duplicates()
display(data.info())

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

71

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: int64(7), object(5)
memory usage: 2.1+ MB


None

**Вывод**

После сведения текстовых значений в столбце с образованием, мы обнаружили 71 дубль в данных (с помощью метода duplicated().sum()) и отметили, что все дубликаты ранее содержали пропуски в столбцах с трудовым стажем и уровнем дохода. При работе с пропусками мы заменили их на среднее значение (доход) и на 0 (стаж). Возможно, заполнение пропусков и привело к появлению дубликатов.


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Удалять дубли можно только после полной обработки данных. Верно! </div>

### Лемматизация

In [5]:
#Импортируем библиотеку для лемматизации
from pymystem3 import Mystem
m = Mystem() 


#применим лемматизацию, чтобы упорядочить значения в столбце с целью получения кредита ('purpose')
#пишем функцию, которая будет брать исходные данные из столбца purpose и лемматизировать их.
def lemmas(purpose):
    purpose = m.lemmatize(purpose)
    return purpose


#определяем столбец, куда будут сводиться результаты лемматизации, полученные в ходе выполнения функции, 
#а также обозначим, из какого столбца берутся данные для выполнения функции.
data['lemmas'] = data['purpose'].apply(lemmas)
    
#проверим результат:
display(data.head())

#Посчитаем частоту упоминаний отдельных лемм для выделения категорий в следующем задании:
from collections import Counter
#объединяем списки, полученные в ходе лемматизации в один список, чтобы использовать Counter
total_lemmas_counter = sum(data['lemmas'], [])
display(Counter(total_lemmas_counter))





children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose                                lemmas  
0               покупка жилья               [покупка,  , жилье, \n]  
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]  
2               покупка жилья               [покупка,  , жилье, \n]  
3  дополнительное образование  [дополнительный,  , образование, \n]  
4             сыграть свадьбу             [сыграть,  , свадьба, \n]

Counter({'покупка': 5897,
         ' ': 33570,
         'жилье': 4460,
         '\n': 21454,
         'приобретение': 461,
         'автомобиль': 4306,
         'дополнительный': 906,
         'образование': 4013,
         'сыграть': 765,
         'свадьба': 2324,
         'операция': 2604,
         'с': 2918,
         'на': 2222,
         'проведение': 768,
         'для': 1289,
         'семья': 638,
         'недвижимость': 6351,
         'коммерческий': 1311,
         'жилой': 1230,
         'строительство': 1878,
         'собственный': 635,
         'подержать': 478,
         'свой': 2230,
         'со': 627,
         'заниматься': 904,
         'сделка': 941,
         'подержанный': 486,
         'получение': 1314,
         'высокий': 1374,
         'профильный': 436,
         'сдача': 651,
         'ремонт': 607})

**Вывод**

Мы выполнили лемматизацию с целью получения списка словоформ в столбце с заявленными целями кредита. В дальнейшем эти леммы можно использовать для кластеризации заявителей по требуемым целям.
Лемматизация проведена с помощью библиотеки pymystem3
Мы использовали функцию, которая принимает на вход значения каждой ячейки из столбца purpose, лемматизирует их и на выход отдает леммы в виде значений в новом столбце нашего датафрейма.
Такой способ позволяет  сэкономить ресурсы, в отличие от циклов.

Далее мы посчитали частоту упоминания отдельных лемм, чтобы понять, какие категории (по целям кредитования) лучше отобрать в следующем задании. Мы использовали метод Counter, предварительно объединив списки с леммами из столбца 'lemmas' в один большой список.

Мы видим, что наиболее часто упоминаются следубщие леммы: 'недвижимость', 'жилье', 'автомобиль', 'образование', 'свадьба'.
Их мы и выделим в качестве словаря в следующем задании.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Нужные категории найдены </div>

### Категоризация данных

In [6]:
#Напишем функцию, которая будет категоризировать заявителей по уровню дохода:
def income_group(total_income):
    if total_income <= 60000:
        return 'ниже среднего'
    if total_income <= 160000:
        return 'средний'
    if total_income <= 250000:
        return 'выше среднего'
    return 'высокий'

#Cоздадим новый столбец, куда написанная функция будет отдавать результат категоризации. 
#Дадим ссылку на столбец, откуда функция будет брать данные на вход
data['income_group'] = data['total_income'].apply(income_group)

#Посмотрим, сколько респондентов в каждой категории по уровню дохода:
display(data['income_group'].value_counts())

#Разделим респондентов на категории, в соответствии с целями получения кредита. Напишем функцию с условиями категоризации:
def purpose_group(lemma):
    if 'автомобиль' in lemma:
        return 'автомобиль'
    if 'образование' in lemma:
        return 'образование'
    if 'свадьба' in lemma:
        return 'свадьба'
    if 'недвижимость' in lemma:
        return 'недвижимость'
    if 'жилье' in lemma:
        return 'недвижимость'
    else:
        return 'другое'

#Создадим новый столбец для результатов категоризации и укажем, откуда брать данные для функции:
data['purpose_group'] = data['lemmas'].apply(purpose_group)

#Посмотрим, сколько респондентов/строк в каждой категории по целям получения кредита:
display(data['purpose_group'].value_counts())

средний          10327
выше среднего     7509
высокий           2812
ниже среднего      806
Name: income_group, dtype: int64

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_group, dtype: int64

**Вывод**

Мы провели категоризацию/кластеризацию заявителей по двум параметрам: по уровню дохода и по цели получения кредита. Это позволит нам сформулировать более точные ответы на вопросы ниже.

Для определения границ каждого сегмента по уровню дохода мы использовали условные значения: ниже 60 тыс.руб., от 60 до 160 тыс.р., от 160 до 250 тыс.р. и выше 250 тыс.р. В качестве значений можно было также использовать квартильные границы, тогда в каждой категории было бы одинаковое количество респондентов. На наш взгляд, целесообразно использовать именно условные границы, т.к. границы квартилей могут изменяться при дополнени/изменении всего массива данных.

Для категоризации по целям получения кредита мы использовали наиболее часто упоминаемы леммы, список которых мы получили в предыдущем задании.

После категоризации мы проверили, сколько респондентов  содержит та или иная категория.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Категоризация поможет нам ответить на вопросы ниже </div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [7]:
#напишем функцию для строки, анализирующую одновременно значения в столбце 'children' и в столбце'debt':
def children_debt(row):
    children = row['children']
    debt = row['debt']
    if children > 0:
        if debt == 1:
            return 'есть просрочка есть дети'
        return 'нет просрочек есть дети'
    if children ==0:
        if debt ==1:
            return 'есть просрочка нет детей'
        return 'нет просрочки нет детей'

#создаем новый столбец для выведения результатов функции:
data['children_debt'] = data.apply(children_debt, axis = 1)

#посмотрим, как распределились значения в новом столбце:
display('Распределение частоты значений в столбце:', data['children_debt'].value_counts())

#Нам необходимо выяснить соотношение числа респондентов с просрочкой по кредитам, к числу респондентов без просрочки, 
#по каждой категории (с детьми/без детей)
#посчитаем долю просрочек от общего числа значений в категориях "с детьми" и "без детей"
display('Доля просрочек у респондентов с детьми - {:.0%}'\
        .format(len(data[data['children_debt'] == 'есть просрочка есть дети'])/len(data[data['children'] != 0])))
display('Доля просрочек у респондентов без детей - {:.0%}'\
        .format(len(data[data['children_debt'] == 'есть просрочка нет детей'])/len(data[data['children'] == 0])))

'Распределение частоты значений в столбце:'

нет просрочки нет детей     13028
нет просрочек есть дети      6685
есть просрочка нет детей     1063
есть просрочка есть дети      678
Name: children_debt, dtype: int64

'Доля просрочек у респондентов с детьми - 9%'

'Доля просрочек у респондентов без детей - 8%'

**Вывод**

Мы написали функцию для строки, чтобы посмотреть на взаимосвязь между наличием детей и возвратом кредита в срок. 
По каждой категории заемщиков мы выяснили количество значений, а потом посчитали соотношение просроченных кредитов с общим числом кредитов в категориях "с детьми" и "без детей".
Мы видим, что у респондентов без детей доля просрочек по кредитам немного ниже: 8% от общего числа против 9% от общго числа респондентов с детьми. Однако эта разница не очень существенна. Таким образом, делаем вывод, что наличие детей не оказывает существенное влияние на вовзрат кредита в срок.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Верно

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [8]:
#напишем функцию для строки, анализирующую одновременно значения в столбце 'family_status_id' и в столбце'debt':
def family_status_debt(row):
    status = row['family_status_id']
    debt = row['debt']
    if debt == 1:
        if status == 0:
            return 'есть просрочка женаты'
        if status == 1:
            return 'есть просрочка гражданский брак'
        if status == 2:
            return 'есть просрочка вдовцы'
        if status == 3:
            return 'есть просрочка в разводе'
        return 'есть просрочка не женаты'
    if debt == 0:
        if status == 0:
            return 'нет просрочки женаты'
        if status == 1:
            return 'нет просрочки гражданский брак'
        if status == 2:
            return 'нет просрочки вдовцы'
        if status == 3:
            return 'нет просрочки в разводе'
        return 'нет просрочки не женаты'

#создаем новый столбец для выведения результатов функции:
data['family_status_debt'] = data.apply(family_status_debt, axis = 1)

#посмотрим, как распределились значения в новом столбце:
display(data['family_status_debt'].value_counts())
#Посчитаем, сколько есть просрочек в % от общего числа респондентов в каждой категории по семейному статусу:
display('Есть просрочка и женаты: {:.0%} от общего числа женатых'\
        .format(len(data[data['family_status_debt'] == 'есть просрочка женаты']) \
                /len(data[data['family_status_id'] == 0])))
display('Есть просрочка и в гражданском браке: {:.0%} от общего числа в гражданском браке'\
        .format(len(data[data['family_status_debt'] == 'есть просрочка гражданский брак']) \
                /len(data[data['family_status_id'] == 1])))
display('Есть просрочка и не женаты: {:.0%} от общего числа неженатых'\
        .format(len(data[data['family_status_debt'] == 'есть просрочка не женаты']) \
                /len(data[data['family_status_id'] == 4])))
display('Есть просрочка вдовцы: {:.0%} от общего числа вдовцов'\
        .format(len(data[data['family_status_debt'] == 'есть просрочка вдовцы']) \
                /len(data[data['family_status_id'] == 2])))
display('Есть просрочка в разводе: {:.0%} от общего числа в разводе'\
        .format(len(data[data['family_status_debt'] == 'есть просрочка в разводе']) \
                /len(data[data['family_status_id'] == 3])))

нет просрочки женаты               11408
нет просрочки гражданский брак      3763
нет просрочки не женаты             2536
нет просрочки в разводе             1110
есть просрочка женаты                931
нет просрочки вдовцы                 896
есть просрочка гражданский брак      388
есть просрочка не женаты             274
есть просрочка в разводе              85
есть просрочка вдовцы                 63
Name: family_status_debt, dtype: int64

'Есть просрочка и женаты: 8% от общего числа женатых'

'Есть просрочка и в гражданском браке: 9% от общего числа в гражданском браке'

'Есть просрочка и не женаты: 10% от общего числа неженатых'

'Есть просрочка вдовцы: 7% от общего числа вдовцов'

'Есть просрочка в разводе: 7% от общего числа в разводе'

**Вывод**


Мы написали функцию для строки, чтобы посмотреть на взаимосвязь между семейным положением и возвратом кредита в срок. 
По каждой категории заемщиков мы выяснили количество значений, а потом посчитали соотношение просроченных кредитов с общим числом кредитов в каждой категории по семейному статусу.
Мы видим, что у респондентов, которые не женаты, доля просрочек по кредитам немного выше: 10% от общего числа.
При этом вдовцы и разведенные, в среднем, более дисциплинированы: только 7% от общего количества респондентов в этих категориях имеют просрочку по кредиту.
Таким образом, делаем вывод, что семейный статус оказывает определенное влияние на возврат кредита в срок. Но возможно, это влияние напрямую связано не со статусом, а с другим фактором, например, с возрастом. Вдовцы и разведенные, как правило, имеют бОльший жизненный опыт, чем неженатые.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [9]:
#напишем функцию для строки, анализирующую одновременно значения в столбце 'income_group' и в столбце'debt':
def income_group_debt(row):
    income_group = row['income_group']
    debt = row['debt']
    if debt == 1:
        if income_group == 'ниже среднего':
            return 'есть просрочка, доход ниже среднего'
        if income_group == 'средний':
            return 'есть просрочка, средний уровень дохода'
        if income_group == 'выше среднего':
            return 'есть просрочка, доход выше среднего'
        return 'есть просрочка, высокий уровень дохода'
    if debt == 0:
        if income_group == 'ниже среднего':
            return 'нет просрочки, доход ниже среднего'
        if income_group == 'средний':
            return 'нет просрочки, средний уровень дохода'
        if income_group == 'выше среднего':
            return 'нет просрочки, доход выше среднего'
        return 'нет просрочки, высокий уровень дохода'

#создаем новый столбец для выведения результатов функции:
data['income_group_debt'] = data.apply(income_group_debt, axis = 1)

#посмотрим, как распределились значения в новом столбце:
display(data['income_group_debt'].value_counts())

#Посчитаем, сколько есть просрочек в % от общего числа респондентов в каждой категории по уровню дохода:
display('Есть просрочка и доход ниже среднего: {:.0%} от общего числа в категории' \
        .format(len(data[data['income_group_debt'] == 'есть просрочка, доход ниже среднего']) \
                /len(data[data['income_group'] == 'ниже среднего'])))
display('Есть просрочка и средний уровень дохода: {:.0%} от общего числа в категории' \
        .format(len(data[data['income_group_debt'] == 'есть просрочка, средний уровень дохода']) \
                /len(data[data['income_group'] == 'средний'])))
display('Есть просрочка и доход выше среднего: {:.0%} от общего числа в категории' \
        .format(len(data[data['income_group_debt'] == 'есть просрочка, доход выше среднего']) \
                /len(data[data['income_group'] == 'выше среднего'])))
display('Есть просрочка и высокий уровень дохода: {:.0%} от общего числа в категории'\
        .format(len(data[data['income_group_debt'] == 'есть просрочка, высокий уровень дохода']) \
                /len(data[data['income_group'] == 'высокий'])))


нет просрочки, средний уровень дохода     9443
нет просрочки, доход выше среднего        6895
нет просрочки, высокий уровень дохода     2618
есть просрочка, средний уровень дохода     884
нет просрочки, доход ниже среднего         757
есть просрочка, доход выше среднего        614
есть просрочка, высокий уровень дохода     194
есть просрочка, доход ниже среднего         49
Name: income_group_debt, dtype: int64

'Есть просрочка и доход ниже среднего: 6% от общего числа в категории'

'Есть просрочка и средний уровень дохода: 9% от общего числа в категории'

'Есть просрочка и доход выше среднего: 8% от общего числа в категории'

'Есть просрочка и высокий уровень дохода: 7% от общего числа в категории'

**Вывод**

Мы написали функцию для строки, чтобы посмотреть на взаимосвязь между уровнем дохода и возвратом кредита в срок. 
По каждой категории заемщиков мы выяснили количество значений, а потом посчитали соотношение просроченных кредитов с общим числом кредитов в каждой категории по уровню дохода.
Мы видим, что у респондентов со средним уровнем дохода доля просрочек по кредитам немного выше: 9% от общего числа.
При этом в группе с доходом ниже среднего только 6% от общего количества респондентов имеют просрочку по кредиту.
Таким образом, делаем вывод, что уровень дохода оказывает определенное влияние на возврат кредита в срок. Но возможно, этот факт связан с определением границ категорий с уровнем дохода: если их сдвинуть, доли просрочек в каждой категории тоже могут измениться.


- Как разные цели кредита влияют на его возврат в срок?

In [10]:
#напишем функцию для строки, анализирующую одновременно значения в столбце 'purpose_group' и в столбце'debt':
def purpose_group_debt(row):
    purpose_group = row['purpose_group']
    debt = row['debt']
    if debt == 1:
        if purpose_group == 'недвижимость':
            return 'есть просрочка, на недвижимость'
        if purpose_group == 'автомобиль':
            return 'есть просрочка, на автомобиль'
        if purpose_group == 'образование':
            return 'есть просрочка, на образование'
        return 'есть просрочка, на свадьбу'
    if debt == 0:
        if purpose_group == 'недвижимость':
            return 'нет просрочки, на недвижимость'
        if purpose_group == 'автомобиль':
            return 'нет просрочки, на автомобиль'
        if purpose_group == 'образование':
            return 'нет просрочки, на образование'
        return 'нет просрочки, на свадьбу'

#создаем новый столбец для выведения результатов функции:
data['purpose_group_debt'] = data.apply(purpose_group_debt, axis = 1)

#посмотрим, как распределились значения в новом столбце:
display(data['purpose_group_debt'].value_counts())

#Посчитаем, сколько есть просрочек в % от общего числа респондентов в каждой категории по цели кредитования:
display('Есть просрочка, на недвижимость: {:.0%} от общего числа в категории' \
        .format(len(data[data['purpose_group_debt'] == 'есть просрочка, на недвижимость']) \
                /len(data[data['purpose_group'] == 'недвижимость'])))
display('Есть просрочка, на автомобиль: {:.0%} от общего числа в категории' \
        .format(len(data[data['purpose_group_debt'] == 'есть просрочка, на автомобиль']) \
                /len(data[data['purpose_group'] == 'автомобиль'])))
display('Есть просрочка, на образование: {:.0%} от общего числа в категории' \
        .format(len(data[data['purpose_group_debt'] == 'есть просрочка, на образование']) \
                /len(data[data['purpose_group'] == 'образование'])))
display('Есть просрочка, на свадьбу: {:.0%} от общего числа в категории'\
        .format(len(data[data['purpose_group_debt'] == 'есть просрочка, на свадьбу']) \
                /len(data[data['purpose_group'] == 'свадьба'])))


нет просрочки, на недвижимость     10029
нет просрочки, на автомобиль        3903
нет просрочки, на образование       3643
нет просрочки, на свадьбу           2138
есть просрочка, на недвижимость      782
есть просрочка, на автомобиль        403
есть просрочка, на образование       370
есть просрочка, на свадьбу           186
Name: purpose_group_debt, dtype: int64

'Есть просрочка, на недвижимость: 7% от общего числа в категории'

'Есть просрочка, на автомобиль: 9% от общего числа в категории'

'Есть просрочка, на образование: 9% от общего числа в категории'

'Есть просрочка, на свадьбу: 8% от общего числа в категории'

**Вывод**

Мы написали функцию для строки, чтобы посмотреть на взаимосвязь между целями кредита и возвратом кредита в срок. 
По каждой категории заемщиков мы выяснили количество значений, а потом посчитали соотношение просроченных кредитов с общим числом кредитов в каждой категории по цели кредитования.
Мы видим, что в среднем, просрочки по кредитам на авто и на образование случаются чаще, чем по кредитам на недвижимость и свадьбу: 9% от общего числа респондентов в категориях 'недвижимость' и 'автомобиль' против 7% и 8% соответственно.
Таким образом, делаем вывод, что цели кредитования оказывают определенное влияние на возврат кредита в срок. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Все так

## Шаг 4. Общий вывод

Мы провели предобработку данных, заменив пропуски, отрицательные значения, удалив дубликаты.
Также выделили укрупненные категории по уровню дохода и целям кредитования, проведя лемматизацию.
Проведя анализ, мы можем заключить, что на возврат кредита в срок с большей вероятностью влияют такие параметры, как:
- семейное положение
- уровень дохода
- цели кредитования

Наличие детей на возврат кредита в срок не оказывает значимого влияния.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Заказчик будет доволен:)

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Спасибо за работу! Проект получился очень сильным. Критичных замечаний нет. Оставила несколько направляющих комментариев. Если у тебя будут вопросы, смело задавай их через куратора. Удачи на следующих этапах!</div>